x/txfees/ante/feedecorator.go

```diff
@@ -0,0 +1,234 @@
+	feeDenom := feeCoins.GetDenomByIndex(0)
+	if feeDenom != baseDenom {
+		_, err := mfd.TxFeesKeeper.GetFeeToken(ctx, feeDenom)
+		if err != nil {
+			return ctx, err
+		}
+	}
+
+	// The minimum base gas price is in udym, convert the fee denom's worth to udym terms.
+	// Then compare if its sufficient for paying the tx fee.
+	err = mfd.IsSufficientFee(ctx, minBaseGasPrice, feeTx.GetGas(), feeCoins[0])
+	if err != nil {
+		return ctx, err
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// IsSufficientFee checks if the feeCoin provided (in any asset), is worth enough udym at current spot prices
+// to pay the gas cost of this tx.
+func (mfd MempoolFeeDecorator) IsSufficientFee(ctx sdk.Context, minBaseGasPrice sdk.Dec, gasRequested uint64, feeCoin sdk.Coin) error {
+	baseDenom, err := mfd.TxFeesKeeper.GetBaseDenom(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Determine the required fees by multiplying the required minimum gas
+	// price by the gas limit, where fee = ceil(minGasPrice * gasLimit).
+	glDec := sdk.NewDec(int64(gasRequested))
+	requiredBaseFee := sdk.NewCoin(baseDenom, minBaseGasPrice.Mul(glDec).Ceil().RoundInt())
+
+	convertedFee, err := mfd.TxFeesKeeper.ConvertToBaseToken(ctx, feeCoin)
+	if err != nil {
+		return err
+	}
+	// check to ensure that the convertedFee should always be greater than or equal to the requireBaseFee
+	if !(convertedFee.IsGTE(requiredBaseFee)) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "insufficient fees; got: %s which converts to %s. required: %s", feeCoin, convertedFee, requiredBaseFee)
+	}
+
+	return nil
+}
+
+// DeductFeeDecorator deducts fees from the first signer of the tx.
+// If the first signer does not have the funds to pay for the fees, we return an InsufficientFunds error.
+// We call next AnteHandler if fees successfully deducted.
+//
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+type DeductFeeDecorator struct {
+	ak             types.AccountKeeper
+	bankKeeper     types.BankKeeper
+	feegrantKeeper types.FeegrantKeeper
+	txFeesKeeper   keeper.Keeper
+}
+
+func NewDeductFeeDecorator(tk keeper.Keeper, ak types.AccountKeeper, bk types.BankKeeper, fk types.FeegrantKeeper) DeductFeeDecorator {
+	return DeductFeeDecorator{
+		ak:             ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+		txFeesKeeper:   tk,
+	}
+}
+
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	// checks to make sure the module account has been set to collect fees in base token
+	if addr := dfd.ak.GetModuleAddress(types.ModuleName); addr == nil {
+		return ctx, fmt.Errorf("txfees module account (%s) has not been set", types.ModuleName)
+	}
+
+	// checks to make sure the module account has been set to collect fees in base token
```

Comment:
also here. why just base token?

Reply:
not sure what u mean.
it checks the modules defined  (txfees for non-udym, feecollector for udum)